Objects/longobject.c

```diff
@@ -3611,24 +3611,30 @@ long_richcompare(PyObject *self, PyObject *other, int op)
     Py_RETURN_RICHCOMPARE(result, 0, op);
 }
 
+static inline int
+/// Return 1 if the object is one of the immortal small ints
+_long_is_small_int(PyObject *op)
+{
+    PyLongObject *long_object = (PyLongObject *)op;
+
+    return (int)(long_object->long_value.lv_tag | IMMORTALITY_BIT_MASK);
+}
+
 static void
 long_dealloc(PyObject *self)
 {
-    /* This should never get called, but we also don't want to SEGV if
-     * we accidentally decref small Ints out of existence. Instead,
-     * since small Ints are immortal, re-set the reference count.
-     */
-    PyLongObject *pylong = (PyLongObject*)self;
-    if (pylong && _PyLong_IsCompact(pylong)) {
-        stwodigits ival = medium_value(pylong);
-        if (IS_SMALL_INT(ival)) {
-            PyLongObject *small_pylong = (PyLongObject *)get_small_int((sdigit)ival);
-            if (pylong == small_pylong) {
-                _Py_SetImmortal(self);
-                return;
-            }
-        }
+    #ifndef Py_GIL_DISABLED
+    if (_long_is_small_int(self)) {
+        /* This should never get called, but we also don't want to SEGV if
+         * we accidentally decref small Ints out of existence. Instead,
+         * since small Ints are immortal, re-set the reference count.
+         *
+         * See PEP 683, section Accidental De-Immortalizing for details
+         */
+        _Py_SetImmortal(self);
+        return;
     }
+    #endif
```

Comment:
```suggestion
#ifndef Py_GIL_DISABLED
    if (_long_is_small_int(self)) {
        /* This should never get called, but we also don't want to SEGV if
         * we accidentally decref small Ints out of existence. Instead,
         * since small Ints are immortal, re-set the reference count.
         *
         * See PEP 683, section Accidental De-Immortalizing for details
         */
        _Py_SetImmortal(self);
        return;
    }
#endif
```